fix(a11y): place textarea between the jaws

bbsmooth opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes locally on my machine.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

Currently in the RWD beta, the `<textarea>` for the monaco editor precedes the upper jaw in the DOM. This has the potential to make navigation confusing for screen reader users as the instructions for the Step in the upper jaw come after the editor, which means they would have to go backwards in the DOM after reading the instructions. Based on the way the page is layed out I think it makes sense to have the instructions come before the editor. Then screen reader users will be able to navigate to the instructions, then to the editor, then to the hint (if needed) in a logical order.

I'm not sure if this is the best way to move the `<textarea>`. Feel free to suggest a better method. I've been testing this for a few days now and I'm not seeing any adverse effects of moving it.

Also, while I was testing I discovered that the NVDA screen reader will skip the `<textarea>` completely in browse mode if it has zero size, so I forced the size to a 1px square.

bbsmooth committed:
fix: place textarea between the jaws

ojeytonwilliams commented:
Quite caveat: I tried navigating the page with Orca, but I'm not particularly familiar with how it's supposed to behave. So, I was curious: how *do* you navigate the DOM directly?  I can get around with the keyboard, but that skips the instructions entirely.

If it's not possible to focus the instructions via keyboard navigation, isn't that the main problem here?

ojeytonwilliams commented:
Brilliant, thanks for taking the time to walk me through that.

So, just to confirm: there's currently no way to modify the accessibility tree without modifying the DOM? The reason I ask is that I'm not at all keen on the idea of directly modifying the editor internals.  That's an easy way to get weird bugs if and when Microsoft updates the editor.

I'm not sure if this has other issues from an a11y perspective, but could we have the instructions in a screen-reader only section before the editor?  Then, if we hide the in-editor instructions, the page order would be instructions, editor, check your code button.

bbsmooth commented:
> there's currently no way to modify the accessibility tree without modifying the DOM

Correct. The accessibility tree is built from the DOM. Bottom line, the DOM is king.

If I understand your idea correctly, instead of moving the `<textarea>`, we duplicate the content in the upper jaw (this includes headings, instructions, and anything else that may be in there) above the monaco editor in the DOM and make it sr-only and then hide the original content from screen readers? I would have to think about that some more, probably test it out. I guess my first response would be could we just move the upper jaw above the monaco editor?

ojeytonwilliams commented:
>  I guess my first response would be could we just move the upper jaw above the monaco editor? Or would that also possibly have repercussions if MS decided to change things down the line?

As in create the jaw above (in the dom) and position it over the editor? I mean... yes in principle, but it's probably quite a bit of work to pull it off.  We'd have to track where over the editor it's supposed to be and make sure it mimics the current behaviour exactly. Not impossible, by any means, but tricky.

bbsmooth committed:
refactor: convert upper jaw to content widget

bbsmooth committed:
fix: upper jaw shift on restart button

bbsmooth committed:
fix: reset height of lower jaw on restart

bbsmooth committed:
fix: upper jaw width resizing

bbsmooth committed:
fix: placement of upper jaw on editor resize

bbsmooth committed:
chore: remove unnecessary code

bbsmooth committed:
chore: add useful comment

bbsmooth commented:
@ojeytonwilliams OK, I changed my tactics and figured out how to use the API provide by monaco to position the upper jaw before the textarea. Instead of using an overlay widget you can use a content widget, which is placed above the textarea. Then just needed to make some changes to position it correctly and viola!

I also fixed a bug with lower jaw not resizing when you restart the step and the hint is showing.

ojeytonwilliams commented on client/src/templates/Challenges/classic/editor.tsx:
```diff
@@ -651,6 +654,10 @@ const Editor = (props: EditorProps): JSX.Element => {
       };
     }
 
+    // Must manually set test feedback height back to zero since
+    // clearTestFeedback does not.
+    const testStatus = document.getElementById('test-status');
+    if (testStatus) testStatus.style.height = '0';
```
Nit: could we create a function for this and replace all the other `testStatus.style.height` resets with it?  It would be nice to have this be the same level of abstraction as `clearTestFeedback`.

ojeytonwilliams reviewed (changes requested):
I really like this solution, @bbsmooth!  Just to satisfy my own curiosity: how did you realise that the content widgets go in the right part of the DOM?

bbsmooth committed:
refactor: create function to set height on test feedback

bbsmooth commented:
@ojeytonwilliams:
> how did you realise that the content widgets go in the right part of the DOM?

Ya, the documentation for monaco is not the greatest, so it wasn't as easy as it should have been. I can't remember exactly what I was googling for, but eventually I came across [an example in their playground](https://microsoft.github.io/monaco-editor/playground.html#interacting-with-the-editor-listening-to-mouse-events) that was using a content widget, and once I knew that such a thing existed and that it put the content above the textarea then everything fell into place.

ojeytonwilliams reviewed (approved):
Yeah, the playground is really helpful.  Way better than the docs, anyway.

Thanks for taking the time to figure out a robust solution, this looks great!

naomi-lgbt reviewed (approved):
This should be good to go!
